File: Largest Divisible Subset/Largest Divisible Subset.py
# Solution for Largest Divisible Subset in PY

class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        nums.sort()
        dp = [1] * n
        prev_idx = [-1] * n
        last_idx, maxL = 0, 1
        for i in range(1, n):
            for j in range(i):
                if nums[i] % nums[j] == 0:
                    if dp[i] < dp[j] + 1:
                        dp[i] = dp[j] + 1
                        prev_idx[i] = j
                    if dp[i] > maxL:
                        maxL = dp[i]
                        last_idx = i
        res = []
        while last_idx != -1:
            res.append(nums[last_idx])
            last_idx = prev_idx[last_idx]
        return res

File: Largest Divisible Subset/Largest Divisible Subset.cpp
// Solution for Largest Divisible Subset in CPP

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<int> dp(n, 1);
        vector<int> prev_idx(n, -1);
        int last_idx = 0;
        int maxL = 1;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[i] < dp[j] + 1) {
                        dp[i] = dp[j] + 1;
                        prev_idx[i] = j;
                    }
                    if (dp[i] > maxL) {
                        maxL = dp[i];
                        last_idx = i;
                    }
                }
            }
        }
        vector<int> res;
        while (last_idx != -1) {
            res.push_back(nums[last_idx]);
            last_idx = prev_idx[last_idx];
        }
        return res;
    }
};
